src/MC/fix_hmc.cpp

```diff
@@ -0,0 +1,1185 @@
+    error->all(FLERR, "fix hmc tried to store incorrect peratom data");
+  }
+  if (stored_peratom_member.address == nullptr) return;
+  int cols;
+  // peratom scalers
+  if (stored_peratom_member.cols == 0) {
+    if (*(T **) current_peratom_member.address != nullptr) {
+      memcpy(*(T **) current_peratom_member.address, stored_peratom_member.address,
+             nlocal * sizeof(T));
+    }
+  } else {
+    // peratom vectors
+    if (stored_peratom_member.cols < 0) {
+      // variable column case
+      cols = *(stored_peratom_member.address_maxcols);
+      *(current_peratom_member.address_maxcols) = *(stored_peratom_member.address_maxcols);
+    } else {
+      // non-variable column case
+      cols = stored_peratom_member.cols;
+    }
+    if (*(T ***) current_peratom_member.address != nullptr) {
+      for (int i = 0; i < nlocal; i++) {
+        memcpy((**(T ***) current_peratom_member.address) + i * cols,
+               (T *) stored_peratom_member.address + i * cols, sizeof(T) * cols);
+      }
+    }
+  }
+  current_peratom_member.cols = stored_peratom_member.cols;
+  current_peratom_member.collength = stored_peratom_member.collength;
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixHMC::setup_arrays_and_pointers()
+{
+  int i, j, m;
+  int pair_flag;
+  int bond_flag;
+  int angle_flag;
+  int dihedral_flag;
+  int improper_flag;
+  int kspace_flag;
+
+  // Per-atom scalar properties to be saved and restored:
+
+  stored_nmax = 0;    // initialize so the memory gets allocated on first save
+  stored_ntotal_body = 0;
+
+  // Determine which energy contributions must be computed:
+  ne = 0;
+  if (force->pair) {
+    pair_flag = 1;
+    ne++;
+  } else
+    pair_flag = 0;
+  if (force->bond) {
+    bond_flag = 1;
+    ne++;
+  } else
+    bond_flag = 0;
+  if (force->angle) {
+    angle_flag = 1;
+    ne++;
+  } else
+    angle_flag = 0;
+  if (force->dihedral) {
+    dihedral_flag = 1;
+    ne++;
+  } else
+    dihedral_flag = 0;
+  if (force->improper) {
+    improper_flag = 1;
+    ne++;
+  } else
+    improper_flag = 0;
+  if (force->kspace) {
+    kspace_flag = 1;
+    ne++;
+  } else
+    kspace_flag = 0;
+
+  // Initialize arrays for managing global energy terms:
+  neg = pair_flag ? ne + 1 : ne;
+  memory->create(eglobal, neg, "fix_hmc:eglobal");
+  delete[] eglobalptr;
+  eglobalptr = new double *[neg];
+  m = 0;
+  if (pair_flag) {
+    eglobalptr[m++] = &force->pair->eng_vdwl;
+    eglobalptr[m++] = &force->pair->eng_coul;
+  }
+  if (bond_flag) eglobalptr[m++] = &force->bond->energy;
+  if (angle_flag) eglobalptr[m++] = &force->angle->energy;
+  if (dihedral_flag) eglobalptr[m++] = &force->dihedral->energy;
+  if (improper_flag) eglobalptr[m++] = &force->improper->energy;
+  if (kspace_flag) eglobalptr[m++] = &force->kspace->energy;
+
+  // Initialize arrays for managing global virial terms:
+  nv = ne;
+  for (j = 0; j < modify->nfix; j++)
```

Comment:
Accessing internal members of the `Modify` class should be avoided for anything but some select core classes. We have added new accessor APIs for those purposes (see issue #2870). So this loop would be replaced by:
```c++
for (const auto &ifix : modify->get_fix_list())
  if (ifix->virial_global_flag) nv++;
```
Same applies to the other loops below.